### PR TITLE
Add new MPUBLISH command

### PR DESCRIPTION
**Pull request by torwig**
The new `MPUBLISH` command allows publishing one or more messages to a channel.
Syntax: `MPUBLISH channel-name message1 message2 ... messageN`

### CODE REVIEW

**Review by enjoy-binbin — COMMENTED** [review]
I'll add some refs first.
redis have this pending issue: [[QUESTION] Why not allow PUBLISH to publish multiple messages?](https://github.com/redis/redis/issues/12213)
and this pending PR: [How MPUBLISH could work #12267](https://github.com/redis/redis/pull/12267)

**Comment by torwig** [comment]
@enjoy-binbin Thank you for the links.

**Comment by torwig** [comment]
I guess I should consider introducing a new command for this instead of changing the `PUBLISH` command.

**Comment by caipengbo** [comment]
> I guess I should consider introducing a new command for this instead of changing the PUBLISH command.

Agree with it

**Comment by enjoy-binbin** [comment]
Do you already want to use this feature? One case is we can pipeline it in this monment (Waiting for Redis to actually implement it if not in a hurry).

or implementing `MPUBLISH channel message1 message2 ...`. This should be possible since it should be implemented in Redis 8 (i see two Redis core-team member agree with it)

**Comment by aleksraiden** [comment]
Thanks, @enjoy-binbin @torwig 

Yes, variant `MPUBLISH channel message1 message2` will be awesome

**Comment by torwig** [comment]
Implemented the feature as a separate command `MPUBLISH`.

**Review by PragmaTwice — APPROVED** [review]

File: src/commands/cmd_pubsub.cc [hunk 1]
```diff
@@ -27,20 +27,48 @@ namespace redis {
...
+ public:
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    int total_receivers = 0;
+
+    for (size_t i = 2; i < args_.size(); i++) {
+      if (!svr->IsSlave()) {
+        redis::PubSub pubsub_db(svr->storage);
+
+        auto s = pubsub_db.Publish(args_[1], args_[i]);
+        if (!s.ok()) {
```
Comment (mapleFU) [inline src/commands/cmd_pubsub.cc:61]: Seems that `Publish` will be likely to success?

File: src/commands/cmd_pubsub.cc [hunk 2]
```diff
@@ -27,20 +27,48 @@ namespace redis {
...
+          return {Status::RedisExecErr, s.ToString()};
+        }
+      }
+
+      int receivers = svr->PublishMessage(args_[1], args_[i]);
+      total_receivers += receivers;
+    }
+
+    *output = redis::Integer(total_receivers);
```
Comment (mapleFU) [inline src/commands/cmd_pubsub.cc:70]: Would an array of int be better here?

Comment (torwig) [reply src/commands/cmd_pubsub.cc:70]: @mapleFU Possibly, yes. However, according to this PR (https://github.com/redis/redis/pull/12267) it's likely that `MPUBLISH` will have a single integer as a response.

**Review by caipengbo — APPROVED** [review]
LGTM

File: src/commands/cmd_pubsub.cc (hunk 2 above)
Comment (mapleFU) [reply src/commands/cmd_pubsub.cc:70]: Oh this LGTM. Seems this is not stable. Maybe we can mark "mpublish" as "experimental" or something, and make it able to change the syntax of it.

Comment (torwig) [reply src/commands/cmd_pubsub.cc:70]: In general, I think it's not easy to utilize the return value of the `PUBLISH` command, except to check whether it is 0 or more. The actual goal of Pub/Sub pattern is to decouple publishers from subscribers and if I want to check the number of receivers I should know the exact number of them. And for example, if I know that there should be 5 subscribers but I received only 4 in response to the `PUBLISH` command, what should I do?

**Review by mapleFU — APPROVED** [review]
This patch general LGTM. I've some tiny problems.

Seems `pubsub_db.Publish` different messages separately. It's ok if all write works well. However, could user find how many message is successed when `Publish` failed?

Another is that this calls a look for each `arg`. Could we make `mpublish` faster, and batch the notifying later? Would this save some time?

File: src/commands/cmd_pubsub.cc (hunk 2 above)
Comment (mapleFU) [reply src/commands/cmd_pubsub.cc:70]: Sounds ok. It's ok to leave it like this

**Review by git-hulk — APPROVED** [review]

**Pull request merged**